Add learning-irregular-verbs-in-english

tanrax opened the pull request:
### Brief summary of what the package does

Helps you learn irregular verbs using the spaced repetition technique.

### Direct link to the package repository

https://github.com/tanrax/lirve.el

### Your association with the package

I am maintainer

### Relevant communications with the upstream package maintainer

None needed

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  You definitely want to align your filename and your function prefixes (lirve is fun and maybe an interesting play on words given that "livre" is book in French).  Otherwise, quick first pass --

`lirve.el` with byte-compile using Emacs 29.1:
```
lirve.el:76:2: Warning: docstring has wrong usage of unescaped single quotes (use \= or different quoting)
In lirve--set-verb-to-learn:
lirve.el:156:124: Warning: reference to free variable `learning-irregular-verbs-in-English--show-translation'
In lirve--show-translation:
lirve.el:187:2: Warning: docstring wider than 80 characters
In end of data:
lirve.el:230:10: Warning: the function `widget-backward' might not be defined at runtime.
lirve.el:190:6: Warning: the function `widget-value-set' might not be defined at runtime.
```

`lirve.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- lirve.el#L136: You can use `not` or `null` (but actually, as for most of these, the truthiness of `(not (eq lirve--widget-field-simple-past nil))` is the same as `lirve--widget-field-simple-past`)
- lirve.el#L140: You can use `not` or `null`
- lirve.el#L232: You can use `not` or `null`
- lirve.el#L233: You can use `not` or `null`
- lirve.el#L234: You can use `not` or `null`
- lirve.el#L235: You can use `not` or `null`
- lirve.el#L236: You can use `not` or `null`
- lirve.el#L237: You can use `not` or `null`
- lirve.el#L278: You can use `not` or `null`
- lirve.el#L280: You can use `not` or `null`
- lirve.el#L282: You can use `not` or `null`
- lirve.el#L156: This double negation can be collapsed (`not` aliases `null`)
- lirve.el#L189: This double negation can be collapsed (`not` aliases `null`)
```

Package and license:
- No .el file matches the name 'learning-irregular-verbs-in-english'
- Can't check Package-Requires if there is no 'main' file
- lirve-verbs.el -- not in package namespace `learning-irregular-verbs-in-english-`
- lirve.el -- not in package namespace `learning-irregular-verbs-in-english-`

tanrax commented:
@riscy First, thank you for your patient. 😁
Second: I updated the code.
I don't know how to fix the following message.
```
lirve.el:230:10: Warning: the function `widget-backward' might not be defined at runtime.
lirve.el:190:6: Warning: the function `widget-value-set' might not be defined at runtime.
```
I read the following references:
https://www.emacswiki.org/emacs/ElispCompilerWarnings
https://stackoverflow.com/questions/8902609/warning-when-byte-compiling-the-function-whitespace-mode-might-not-be-defined
https://emacs.stackexchange.com/questions/34319/function-might-not-be-defined-at-runtim
I think the best solution is to use `with-no-warnings`. What do you think?
PS: For now I want it to work, maybe later I will think about changing the name. 😂

riscy commented:
Hi again - thanks for the update.  Please make sure your recipe is correct - I think the repo is https://github.com/tanrax/lirve.el and the package is `lirve` but your recipe is pointing to something else.

>  I think the best solution is to use with-no-warnings. What do you think?

No, you probably want to just `require` the dependencies where these are defined.  I can take a closer look after your recipe is fixed. :)

tanrax commented:
@riscy Yes, please. I want to learn for you! 
I updated the repo address.

riscy commented:
Looks good - thanks.  Here's the latest --

`lirve.el` with byte-compile using Emacs 29.1:
```
lirve.el:231:10: Warning: the function `widget-backward' might not be defined at runtime.
lirve.el:191:6: Warning: the function `widget-value-set' might not be defined at runtime.
```
`widget-backward` and `widget-value-set` are not autoloaded and so your `(eval-when-compile (require 'wid-edit))` will only guarantee that these are loaded at compile time, not load time.  Just write `(require 'wid-edit)` instead to solve this.

Package and license:
- No .el file matches the name 'learning-irregular-verbs-in-english' - your recipe should be `(lirve :repo "tanrax/lirve.el" :fetcher github)`

tanrax commented:
I have made the changes. Thanks!

riscy commented on recipes/lirve:
```diff
@@ -0,0 +1 @@
+(learning-irregular-verbs-in-english :repo "tanrax/lirve.el" :fetcher github)
```
```suggestion
(lirve :repo "tanrax/lirve.el" :fetcher github)
```

riscy reviewed (commented):
Thanks - we need one more update to the recipe (see suggestion).  Here's the latest --

`lirve.el` with package-lint 20240220.2303 and `package-lint-main-file` = "lirve.el":
```
1 issue found:
372:3: error: You should depend on (emacs "26.1") if you need `display-line-numbers-mode'.
```

tanrax commented:
> Thanks - we need one more update to the recipe (see suggestion). Here's the latest --
> 
> `lirve.el` with package-lint 20240220.2303 and `package-lint-main-file` = "lirve.el":
> 
> ```
> 1 issue found:
> 372:3: error: You should depend on (emacs "26.1") if you need `display-line-numbers-mode'.
> ```

I have uploaded the required version from 25.1 to 26.1

riscy commented:
Thanks for the updates - checklist looks clear, recipe looks good.

This looks like it is focused on a Spanish audience, based on the contents of lirve-verbe.el.  I think you should definitely advertise this in your summary line which is currently:
```elisp
;;; lirve.el --- Learn irregular verbs in English. -*- lexical-binding: t
```
And while I'm on metadata, your `;;; Commentary` for lirve-verbs.el looks truncated:
```elisp
;; This file contains a list of irregular verbs in English, with their
```

The rest of these are minor but may help with style --

You don't need a `progn` for the "else" part of an `if`, e.g. in `lirve--toggle-layout-finish`.  You also don't need a `progn` for a `when` form, e.g. in `lirve--save-verb-unresolved`.  And note that instead of `(when (not ...` you can use `unless` (but some people don't like `unless`; up to you).

You're using numbers to code state, i.e. `(defvar lirve--state 1) ;; 1: lirve--start, 2: playing (before first check), 3: win (show success layout)` but you could equally easily use [keyword symbols](https://www.gnu.org/software/emacs/manual/html_node/elisp/Symbol-Type.html) and have much more readable code.  E.g. `(defvar lirve--state :start)`

You have a variable called `lirve--is-resolve` but lisps have a nice convention of indicating predicates as e.g. `lirve--resolved-p`.  It's good to get used to this convention because you'll see it elsewhere.

You should be using `defconst` in addition to `defvar` to keep your constants and variables clear. For example `lirve--text-success` will never change; use `defconst` instead of `defvar`.

`lirve--kill-app` is `(interactive)`, but private because its prefix is `livre--`.  Anything interactive should be public, because it *will* be "publicized" through e.g. `M-x` autocompletes.

tanrax commented:
Thanks @riscy for the comments and the effort in reviewing the code. I'm going to answer by points.
1. I have included constants as you have suggested. It's true, there are variables that are actually constants.
2. I have renamed the function lirve--is-resolve to lirve--resolved-p , and lirve--kill-app to lirve-kill-app .
3. The application is not for people who speak Spanish, only for those who want to learn English. The translation is only activated by configuration, and there is the possibility of adding other languages. In the README I leave an example for this.
4. I changed these numbers with keyword symbols. Visually it is much clearer.
5. I have fixed the truncated comment.
Thanks for the feedback!

tanrax commented:
Another update: Replace (when (not ...)) with (unless ...) https://github.com/tanrax/lirve.el/commit/2a6f32f73d5563fc37a3f662649b3dd99425d16c

riscy commented:
> The application is not for people who speak Spanish, only for those who want to learn English [...] In the README I leave an example for this.

Your README says: `Only available in Spanish (at the moment)` and `If you want to add more languages, make a PR with the translations`.  Any package can be anything if people make a PR - your metadata should tell people what your package is today.  Moreover, users who install this package won't have access to this README, and the MELPA index won't feature your README either, just the contents of your `;;; Commentary`.  

> Could you give me an example without progn?

Sure, here's your current function:
```
(defun lirve--save-verb-unresolved (infinitive)
  "Save the verb unresolved to `lirve--verbs-unresolved' and to the file.
Argument INFINITIVE is verb to learn."
  (when infinitive
    (progn
      (setq lirve--verbs-unresolved (delete-dups (append lirve--verbs-unresolved (list infinitive))))
     (with-temp-file (lirve--full-path-unresolved)
       (prin1 lirve--verbs-unresolved (current-buffer))))))
```
and here it is without progn:
```elisp
(defun lirve--save-verb-unresolved (infinitive)
  "Save the verb unresolved to `lirve--verbs-unresolved' and to the file.
Argument INFINITIVE is verb to learn."
  (when infinitive
    (setq lirve--verbs-unresolved (delete-dups (append lirve--verbs-unresolved (list infinitive))))
    (with-temp-file (lirve--full-path-unresolved)
      (prin1 lirve--verbs-unresolved (current-buffer)))))
```
Here's another one:
```

(defun lirve--toggle-layout-finish ()
  "Toggle the layout to success."
  (if (eq lirve--state :win)
      (progn
	;; Show translate
	(lirve--show-translation)
	;; Cursor to end
	(goto-char (point-max))
	;; Remove check button
	(widget-delete lirve--widget-button-check)
	(setq lirve--widget-button-check nil)
	;; Remove space after check button
	(widget-delete lirve--widget-item-space-before-check)
	(setq lirve--widget-item-space-before-check nil)
	;; Remove show solution button
	(widget-delete lirve--widget-button-show-solution)
	(setq lirve--widget-button-show-solution nil)
	;; Text success
	(setq lirve--widget-item-space-before-success (widget-create 'item
								    ""))
	(setq lirve--widget-message-success (widget-create 'item
							  (if lirve--resolved-p lirve--text-success lirve--text-fail)))
	(setq lirve--widget-item-space-after-success (widget-create 'item
								   "\n"))
	;; Lirve--Replay button
	(setq lirve--widget-button-lirve--replay (widget-create 'push-button
							      :size 20
							      :notify (lambda (&rest ignore)
									(lirve--replay))
							      lirve--text-button-lirve--replay))
	;; Space
	(setq lirve--widget-item-space-between-buttons (widget-create 'item
								     "\n"))
	;; Quit button
	(setq lirve--widget-button-quit (widget-create 'push-button
						      :size 20
						      :notify (lambda (&rest ignore) (lirve-kill-app))
						      lirve--text-button-quit))
	(widget-backward 2))
    (progn
      (when lirve--widget-item-space-before-success (widget-delete lirve--widget-item-space-before-success))
      (when lirve--widget-message-success (widget-delete lirve--widget-message-success))
      (when lirve--widget-item-space-after-success (widget-delete lirve--widget-item-space-after-success))
      (when lirve--widget-button-lirve--replay (widget-delete lirve--widget-button-lirve--replay))
      (when lirve--widget-item-space-between-buttons (widget-delete lirve--widget-item-space-between-buttons))
      (when lirve--widget-button-quit (widget-delete lirve--widget-button-quit)))))
```
...and without progn:
```elisp

(defun lirve--toggle-layout-finish ()
  "Toggle the layout to success."
  (if (eq lirve--state :win)
      (progn
	;; Show translate
	(lirve--show-translation)
	;; Cursor to end
	(goto-char (point-max))
	;; Remove check button
	(widget-delete lirve--widget-button-check)
	(setq lirve--widget-button-check nil)
	;; Remove space after check button
	(widget-delete lirve--widget-item-space-before-check)
	(setq lirve--widget-item-space-before-check nil)
	;; Remove show solution button
	(widget-delete lirve--widget-button-show-solution)
	(setq lirve--widget-button-show-solution nil)
	;; Text success
	(setq lirve--widget-item-space-before-success (widget-create 'item
								    ""))
	(setq lirve--widget-message-success (widget-create 'item
							  (if lirve--resolved-p lirve--text-success lirve--text-fail)))
	(setq lirve--widget-item-space-after-success (widget-create 'item
								   "\n"))
	;; Lirve--Replay button
	(setq lirve--widget-button-lirve--replay (widget-create 'push-button
							      :size 20
							      :notify (lambda (&rest ignore)
									(lirve--replay))
							      lirve--text-button-lirve--replay))
	;; Space
	(setq lirve--widget-item-space-between-buttons (widget-create 'item
								     "\n"))
	;; Quit button
	(setq lirve--widget-button-quit (widget-create 'push-button
						      :size 20
						      :notify (lambda (&rest ignore) (lirve-kill-app))
						      lirve--text-button-quit))
	(widget-backward 2))
    (when lirve--widget-item-space-before-success (widget-delete lirve--widget-item-space-before-success))
    (when lirve--widget-message-success (widget-delete lirve--widget-message-success))
    (when lirve--widget-item-space-after-success (widget-delete lirve--widget-item-space-after-success))
    (when lirve--widget-button-lirve--replay (widget-delete lirve--widget-button-lirve--replay))
    (when lirve--widget-item-space-between-buttons (widget-delete lirve--widget-item-space-between-buttons))
    (when lirve--widget-button-quit (widget-delete lirve--widget-button-quit))))
```

tanrax commented:
@riscy Thank for the feedback.
I just uploaded a new version updating `;;; Commentary` and deleting some `progn`.

riscy commented:
Thank you!
